data/transactions/logic/opcodes.go

```diff
@@ -359,11 +359,117 @@ type typedList struct {
 	Effects string
 }
 
+// StackChangeExplanation explains the effect of an opcode over the stack.
+// An opcode may delete a few variables from stack, then add a few to stack.
+type StackChangeExplanation struct {
+	Deletions int
+	Additions int
+}
+
+type debugStackExplain func(*EvalContext) StackChangeExplanation
+
 // Proto describes the "stack behavior" of an opcode, what it pops as arguments
 // and pushes onto the stack as return values.
 type Proto struct {
 	Arg    typedList // what gets popped from the stack
 	Return typedList // what gets pushed to the stack
+
+	// Explain is the pointer to the function used in debugging process during simulation:
+	// - on default construction, Explain relies on Arg and Return count.
+	// - otherwise, we need to explicitly infer from EvalContext, by registering through explain function
+	Explain debugStackExplain
+}
+
+func (p Proto) stackExplain(e debugStackExplain) Proto {
+	p.Explain = e
+	return p
+}
+
+func defaultDebugExplain(argCount, retCount int) debugStackExplain {
+	return func(_ *EvalContext) StackChangeExplanation {
+		return StackChangeExplanation{Deletions: argCount, Additions: retCount}
+	}
+}
+
+// NextStackChange is a helper function that queries EvalContext for the coming stack change of the current PC.
+func (cx *EvalContext) NextStackChange() StackChangeExplanation {
+	return (opsByOpcode[cx.version][cx.program[cx.pc]].Explain)(cx)
+}
+
+func opReturnStackChange(cx *EvalContext) StackChangeExplanation {
+	return StackChangeExplanation{Deletions: len(cx.Stack), Additions: 1}
```

Comment:
Interesting, so return removes the whole stack and re-adds the last item?

At a technical level, this makes sense, but I wonder if reporting no stack change would be simpler?

Reply:
Never mind, I actually like this behavior now